### ISSUE TITLE
GitGutter runs twice when opening file.

### ISSUE THREAD

**Issue by deathaxe** [jisaacks/GitGutter#329]
Hello guys,

GitGutter is setup as follows by default.

```js
"live_mode": true,
"focus_change_mode": true,
```

That's why GitGutter is triggerd with `on_load_async` and `on_activate_async` each time a file is opened. After some testing I would suggest the following little code change to avoid running git and diff twice per file.

```python
class GitGutterEvents(sublime_plugin.EventListener):
   # ...
    def on_activated(self, view):
        if self.focus_change_mode() and not self.live_mode():
            self.debounce(view, 'activated')
```

**Comment by rchl** [comment]
That would break updating on focus when focusing file that was already open.
I agree that is a bug or possible enhancement at least.

**Comment by deathaxe** [comment]
This was my first concern, too, but it does not.

I added some print(..) to the `on_...` event callbacks and did the following steps to check, what's happening.

1. Open a file "setup.py" from another git repo in Sublime Text. 
2. Do some changes to see GitGutter running smoothly.
3. Leave Sublime Text and open the same file in another editor and do some changes.
4. Return to Sublime Text. As soon as the view gets focus the file gets reloaded and `on_load_async` is triggered. This works if the view is already in the foreground or if it is brought back to foreground with tab change.

So if the file is modified externally, it is reloaded as soon as the view gets focused by the reloading mechanism of Sublime Text. Any other change is done by user, which is handled by `on_modified`.

I'll do some further investigation but can't see a situation this code change could break GitGutter.

**Comment by rchl** [comment]
But it would be a problem on starting sublime with files already open. Those files won't get on_load event and then focusing them would not trigger update.

**Comment by deathaxe** [comment]
I don't see on_activate either. No matter if I try with my "fix" from above or not, if a file is opened by Sublime Text (e.g.: `sublime_text.ext setup.py`) it doesn't show gutter icons at all. This seems related with [core issues #1457](https://github.com/SublimeTextIssues/Core/issues/1457) as Sublime Text does not wait for all plugins being loaded until redirecting plugin commands or events.

In my case GitGutter loads too late to see the events at all. This could only fixed by a poll in the `plugin_loaded()` function.

**Comment by rchl** [comment]
The missing `on_activated` for initially open view is an existing bug that can be fixed by just forcibly calling `on_activated` method for active view from `plugin_loaded()`. But that is a problem only for initially open first. Focusing views after that will work fine because `on_activated` will trigger.

Your second is probably not ideal either as technically it's possible to load the file without it being activated. Whether we want or not to run Git Gutter in that case is maybe not entirely clear but it would be safer to run it.

**Comment by rchl** [comment]
This can probably be fixed easily on the GitGutterHandler level now. Checking if there is pending promise and dropping the request if so. That would be a generic way to fix this and potentially other similar cases where we trigger git too often.

**Comment by deathaxe** [comment]
> The missing on_activated for initially open view is an existing bug ...

I agree with you and already thought about creating an core issue for it - if not already exists.

> ... technically it's possible to load the file without it being activated.

I have even been looking for ways to avoid GitGutter from running in those situations. If a user runs "find an replace in files", which results in hundreds of files being opened and changed, git and diff is called for each of them concurrently. This may take a while and cost a lot of CPU and RAM and is absolutely unnecessary. The gutter needs to be updated only for visible views.

If a file like the mensioned `setup.py` is displayed in the first column and a second column is focused when leaving Sublime Text to change `setup.py` externally, it receives an `on_activate` as soon as Sublime Text is focused again. So it gets updated correctly even though it is not the active view at this moment.

Opening Sublime Text with a single file or adding a file to an existing window always results in the file being activated. If not it might not be visible and therefore GitGutter can ignore it. As soon as the file is activated it gets diffed.

The minimap as an example is repainted on the first scroll or modify event anyway, no matter if gutter was updated or not.

**Comment by deathaxe** [comment]
Fixed with PR #331
